teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,538 @@
+
+It's also beneficial that the schemas defined in Chains can follow the field requirements in in-toto attestations. Currently, the type hinting is following a readable formats that users will find it more readable, such as results being populated in [upload-pypi](https://github.com/tektoncd/catalog/blob/f4708d478ee8fac6b5b68347cde087cb7c1d1b1c/task/upload-pypi/0.1/upload-pypi.yaml) and [jib-maven](https://github.com/tektoncd/catalog/blob/6a6f3543fa14d7d840fd13d19ba4452e5e319830/task/jib-maven/0.4/jib-maven.yaml) Catalog Tasks, but these gives difficulties about standarizing signable target schemas, as each signable target's metadata can be faily different from each other. For example, a Git source usually needs a git commit and revision, which are super different from what is needed for OCI images, which need image url and its digest. In-toto attestations provide a way to unify the identifiers for multiple types of signable artifacts, which is described in the [Use Cases](#use-cases).
+
+Generating the objects that satisfy these schemas, from human readable provenance metadata can be done by writing provenence data using Steps in Tasks into these schema structures in TaskRuns and populate them PipelineRuns' Results, and there is also a discussion around adding a new field `provenance` in PipelineRun and TaskRun, which only allows [Trusted Resources](https://github.com/tektoncd/community/blob/2413ad70a742a6e9103c531e1c5788b2b392a7eb/teps/0091-verified-remote-resources.md#requirements) to generate provenance info in this field. Chains should be able to accomodate what is currently capable in TaskRun and PipelineRun results, and other potential new fields specifically for provenances.
+
+### Notes and Caveats
+
+When these well defined structures grow as richer provenance are pursued, the size of the structure can eventually grow beyond the limit of the container termination message. [TEP 0086](https://github.com/tektoncd/community/pull/521) is addressing this issue.
+
+## Design Details
+
+### Signable Artifacts
+
+All signable artifacts should be provided in a structure that is well defined in the proposal. If any field is missed, the signable artifacts will be skipped and an error will be thrown. The users need to follow the naming pattern in order the result objects can be captured.
+
+As described above, users of Tekton Chains need to have two results, `ARTIFACT_INPUTS` and `ARTIFACT_OUTPUTS`, that are arrays of strings whichs stores the signable artifacts' result names:
+
+```yaml
+results:
+  - name: ARTIFACT_INPUTS
+    value: ["ARTIFACT-INPUT-NAME-1", "ARTIFACT-INPUT-NAME-1"]
+  - name: ARTIFACT_OUTPUTS
+    value: ["ARTIFACT-OUTPUT-NAME-1"]
```

Comment:
There are multiple threads such as  #alternatives section in the TEP and https://github.com/tektoncd/community/blob/main/teps/0082-workspace-hinting.md about using a separate field for providing provenance metadata together with Trusted Resources, and the down sides for using results as future field for such purpose is not ideal. So I'm concerned that if I propose to add "signable" field in results, it may not be a future solution that we will be sticking to.

Reply:
Another thing along this route could be - adding an extra key in the object, such as "artifact_type"? If the results are having this key, and the value is either "input" or "output", then this is a result that Chains is looking for. This doesn't need us to add a field in results in Pipeline.

Reply:
I prefer this over requiring task authors to plumb through result names, though a few caveats:

Putting data in object keys means that we're breaking from the intoto spec - e.g. what happens if a user just outputs a vanilla intoto material? Is intoto compatibility a goal for this proposal (it feels like this is what we're aiming for, but it's not explicitly called out as a goal)?

By putting the field in the result spec, this has the nice property of not having to need to modify the payload the Task needs to output (e.g. it can be a vanilla intoto spec), but if we'll need a spec change it might make more sense to just move forward with the status-based provenance.

Reply:
I think having the schema being compatible with in-toto provenance should be one of the goals. I also agree that adding a field in results, which requires spec change, may not satisfy our future needs. That said, maybe keeping the results to follow naming type hinting may still be the temporary pattern we should follow before the status-based provenance being landed. WDYT @wlynch @afrittoli ?

Reply:
I can live with it, though I think we should change the arrays - I'm not sure if there's an easy way to have the Task provide these without hardcoding them. I don't think there's a way to derive the values from the [replacement context](https://tekton.dev/docs/pipelines/variables/#variables-available-in-a-task).

Reply:
> 

Do you mean for Task authors how they can easily add those suffices?